File: src/main/scala/universalelectricity/core/grid/TickingGrid.scala
package universalelectricity.core.grid

import universalelectricity.api.core.grid.IUpdate

import scala.collection.convert.wrapAll._
import scala.reflect.ClassTag

/**
 * A grid that ticks. May be multi-threaded.
 *
 * @tparam N - The type of the node
 *
 * @author Calclavia
 */
class TickingGrid[N <: Node[N] : ClassTag] extends NodeGrid[N] with IUpdate
{
  /** Upon init, add this grid into the ticker. */
  UpdateTicker.addUpdater(this)

  override def add(node: N)
  {
    UpdateTicker.enqueue(() => super.add(node))
  }

  override def remove(node: N)
  {
    UpdateTicker.enqueue(() => super.remove(node))
  }

  /**
   * An grid update called only server side.
   */
  def update(deltaTime: Double)
  {
    /**
     * TODO: Perform test to check if parallel evaluation is worth it.
     */
    getNodes().par.foreach(_.update(deltaTime))
  }

  def canUpdate() = getNodes().size > 0

  def continueUpdate = canUpdate()

}

File: src/main/scala/universalelectricity/core/grid/TickingGrid.java
package universalelectricity.core.grid;

import scala.reflect.ClassTag;
import universalelectricity.api.core.grid.IUpdate;

/**
 * Created by robert on 8/11/2014.
 */
public class TickingGrid<N extends Node> extends NodeGrid<N> implements IUpdate {

    public TickingGrid(Class node)
    {
        super(node);
        UpdateTicker.addUpdater(this);
    }

    @Override
    public void update(double deltaTime)
    {
        for(N node : this.getNodes())
        {
            node.update(deltaTime);
        }
    }

    @Override
    public boolean canUpdate() {return getNodes().size() > 0; }

    @Override
    public boolean continueUpdate() { return canUpdate();}
}
